helix-core/src/selection.rs

```diff
@@ -363,7 +363,9 @@ impl Selection {
     /// Adds a new range to the selection and makes it the primary range.
     pub fn remove(mut self, index: usize) -> Self {
         self.ranges.remove(index);
-        if index < self.primary_index || self.primary_index == self.ranges.len() {
+        if index < self.primary_index
+            || self.primary_index == self.ranges.len() && self.primary_index > 0
+        {
```

Comment:
It's not actually valid to remove all the ranges from a selection. I think rather than do this, there should be an `assert!(self.ranges.len() > 1)` to catch this

Reply:
To fix the search function where we `remove` then `push`, let's add a `replace(index: usize, range: Range) -> Range` that returns the original range.

```rust
pub fn replace(mut self, index: usize, range: Range) -> Range {
  let original = std::mem::replace(&mut self.ranges[index], range);
  self.normalize();
  original
}
```